lib/src/rules/use_late_for_private_fields_and_variables.dart

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/features.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+
+import '../analyzer.dart';
+import '../util/dart_type_utilities.dart';
+
+const _desc = r'Use late for private members with non-nullable type.';
+
+const _details = r'''
+
+Use late for private members with non-nullable type that are always expected to
+be non-null. Thus it's clear that the field is not expected to be `null` and it
+avoids null checks.
+
+**BAD:**
+```
+int? _i;
+m() {
+  _i!.abs();
+}
+```
+
+**GOOD:**
+```
+late int _i;
+m() {
+  _i.abs();
+}
+```
+
+''';
+
+class UseLateForPrivateFieldsAndVariables extends LintRule
+    implements NodeLintRule {
+  UseLateForPrivateFieldsAndVariables()
+      : super(
+          name: 'use_late_for_private_fields_and_variables',
+          description: _desc,
+          details: _details,
+          maturity: Maturity.experimental,
+          group: Group.style,
+        );
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends UnifyingAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  CompilationUnitElement currentUnit;
+
+  static final lateables =
+      <CompilationUnitElement, List<VariableDeclaration>>{};
+  static final nullableAccess = <CompilationUnitElement, Set<Element>>{};
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    if (node.featureSet.isEnabled(Feature.non_nullable)) {
+      currentUnit = node.declaredElement;
+      lateables.putIfAbsent(currentUnit, () => []);
+      nullableAccess.putIfAbsent(currentUnit, () => {});
+
+      super.visitCompilationUnit(node);
+
+      final units = node.declaredElement.library.units;
+      if (units.every(lateables.keys.contains)) {
```

Comment:
I'm not positive that there's an issue here, but it might be the case that generated files won't be visited, which would mean that libraries with generated parts wouldn't ever satisfy this condition. (And if it's the case today that all parts are always visited, there's no reason to think that they always will be.)

I guess what I was thinking about was more along the lines of
```
if (node is a library and nnbd is enabled for the library) {
  for (each unit in `LinterContext.allUnits`)
    visit the unit to gather lateables (in per unit caches) and nullableAccesses (in a single collection)
  for (each unit in `LinterContext.allUnits`)
    create lints for any lateable that isn't referenced in a nullable access
```